File: tests/test_kernels.c
#include <stddef.h>
#include <math.h>
#include <stdlib.h>
#include "test_kernels.h"

double dot(const double *restrict x, const double *restrict y, int len_x)
/* Return dot product between x and y

Contract: len_x=len(x)
*/
{
    double acc = 0.0;
    for (int i = 0; i < len_x; ++i)
        acc += x[i] * y[i];
    return acc;
}

void scale(double *restrict x, double alpha, int len_x, double *restrict out_x)
/* Scale vector x by alpha and store the result in out

Contract: len_x=len(x); len(out_x)=len_x;
*/
{
    for (int i = 0; i < len_x; ++i)
        out_x[i] = alpha * x[i];
}

void cross(const double *restrict a, const double *restrict b, double *restrict out)
/* Compute the cross product between 3D vectors a and b, store the result in out 

Contract: len(out)=3;
*/

{
    out[0] = a[1] * b[2] - a[2] * b[1];
    out[1] = a[2] * b[0] - a[0] * b[2];
    out[2] = a[0] * b[1] - a[1] * b[0];
}

double complex_magnitude(const ComplexPair *z)
/* Return sqrt(re^2 + im^2) */
{
    return sqrt(z->real * z->real + z->imag * z->imag);
}

double kinetic_energy(const Particle *p, int len_p)
/* Sum 0.5*|v|^2 over particles

Contract: len_p=len(p);
*/
{
    double sum = 0.0;
    for (int i = 0; i < len_p; ++i) {
        double vx = p[i].vel[0];
        double vy = p[i].vel[1];
        double vz = p[i].vel[2];
        sum += 0.5 * (vx * vx + vy * vy + vz * vz);
    }
    return sum;
}


void split_vectors(const double *restrict inp, int len_inp, double *restrict out1, double *restrict out2)
/* Split even/odd elements of inp into out1/out2

Contract: len_inp=len(inp); len(out1)=len_inp/2; len(out2)=len_inp/2;
*/
{
    for (int i = 0; i < len_inp; ++i) {
        if (i % 2 == 0)
            out1[i / 2] = inp[i];
        else
            out2[i / 2] = inp[i];
    }
}

void make_particles(double speed, Particle *out_p, int len_p)
/* Fill particles with unit positions and uniform speed on x

Contract: len(out_p)=len_p;
*/
{
    for (int i = 0; i < len_p; ++i) {
        out_p[i].pos[0] = 1.0;
        out_p[i].pos[1] = 1.0;
        out_p[i].pos[2] = 1.0;
        out_p[i].vel[0] = speed;
        out_p[i].vel[1] = 0.0;
        out_p[i].vel[2] = 0.0;
    }
}


double geom2d_norm(double x, double y)
/* Get the norm of a 2D vector */
{
    return sqrt(x*x + y*y);
}

void merge_sorted(const double *restrict a, const double *restrict b, int len_a, int len_b, double *out, int *out_len)
/* Return array with not repeated ascending values from a and b that are assumed to be sorted

Contract: len_a=len(a); len_b=len(b); len(out)=len_a+len_b;
Post-Contract: len(out)=out_len;
*/
{
    int ia = 0, ib = 0, k = 0;
    double last = NAN;
    while (ia < len_a && ib < len_b) {
        double va = a[ia];
        double vb = b[ib];
        double v;
        if (va < vb) {
            v = va;
            ia++;
        } else if (vb < va) {
            v = vb;
            ib++;
        } else {
            v = va;
            ia++;
            ib++;
        }
        if (k == 0 || v != last) {
            out[k++] = v;
            last = v;
        }
    }
    while (ia < len_a) {
        double v = a[ia++];
        if (k == 0 || v != last) {
            out[k++] = v;
            last = v;
        }
    }
    while (ib < len_b) {
        double v = b[ib++];
        if (k == 0 || v != last) {
            out[k++] = v;
            last = v;
        }
    }
    *out_len = k;
}

double *alloc_random_array(int *out_len)
/* Allocate an array with random-ish length

Own: return
Contract: len(return)=out_len
*/
{
    int n = (int)(rand() % 5 + 3); /* length between 3 and 7 */
    double *arr = (double *)malloc(n * sizeof(double));
    for (int i = 0; i < n; ++i)
        arr[i] = (double)(i + 1);
    *out_len = n;
    return arr;
}

File: tests/test_kernels.py
from pathlib import Path

import numpy as np

import pytest

from tinycwrap import CModule


@pytest.fixture(scope="module")
def cm():
    c_path = Path(__file__).resolve().parent / "test_kernels.c"
    return CModule(c_path)


def test_dot(cm):
    x = np.array([1.0, 2.0, 3.0], dtype=np.float64)
    y = np.array([4.0, 5.0, 6.0], dtype=np.float64)
    assert cm.dot(x, y, len_x=len(x)) == np.dot(x, y)


def test_scale_auto_output(cm):
    x = np.arange(10, dtype=np.float64)
    scaled_auto = cm.scale(x, 1.1, len_x=len(x))
    np.testing.assert_allclose(scaled_auto, x * 1.1)


def test_scale_explicit_output(cm):
    x = np.arange(10, dtype=np.float64)
    out = np.empty_like(x)
    scaled_explicit = cm.scale(x, 2.0, len_x=len(x), out_x=out)
    np.testing.assert_allclose(out, x * 2.0)
    np.testing.assert_allclose(scaled_explicit, x * 2.0)


def test_mat_add_shape_contract(cm):
    a = np.arange(6, dtype=np.float64).reshape(2, 3)
    b = np.ones_like(a)
    out = cm.mat_add(a, b)
    assert out.shape == (2, 3)
    np.testing.assert_allclose(out, a + b)

    flat = np.zeros(a.size, dtype=np.float64)
    reshaped = cm.mat_add(a, b, out=flat)
    assert reshaped.shape == (2, 3)
    np.testing.assert_allclose(reshaped, a + b)
    np.testing.assert_allclose(flat.reshape(a.shape), a + b)


def test_struct_wrapper(cm):
    cp = cm.ComplexPair(real=1.5, imag=-2.5)
    assert repr(cp) == "ComplexPair(real=1.5, imag=-2.5)"
    assert cp.real == 1.5
    assert cp.imag == -2.5
    cp.real = 3.0
    cp.imag = 4.0
    assert cp.real == 3.0
    assert cp.imag == 4.0
    np.testing.assert_equal(cp.dtype.fields.keys(), {"real", "imag"})
    assert cp.dtype["real"] == np.dtype(np.float64)


def test_struct_argument(cm):
    cp = cm.ComplexPair(real=3.0, imag=4.0)
    mag_sq = cm.complex_magnitude(cp)
    assert np.isclose(mag_sq, 5.0)


def test_struct_pointer_inplace_no_return(cm):
    cp = cm.ComplexPair(real=2.0, imag=-3.0)
    res = cm.scale_complex(cp, 2.0)
    assert res is None
    assert cp.real == 4.0
    assert cp.imag == -6.0


def test_struct_scalar_output_returns_struct_object(cm):
    a = cm.ComplexPair(real=2.0, imag=4.0)
    b = cm.ComplexPair(real=6.0, imag=10.0)
    mid = cm.midpoint_complex(a, b)
    assert isinstance(mid, cm.ComplexPair)
    assert mid.real == 4.0
    assert mid.imag == 7.0

    out = cm.ComplexPair()
    res = cm.midpoint_complex(a, b, out_z=out)
    assert res is None
    assert out.real == 4.0
    assert out.imag == 7.0


def test_struct_array_output_returns_array_wrapper(cm):
    pairs = cm.make_complex_pairs()
    assert isinstance(pairs, cm.ComplexPairArray)
    assert repr(pairs) == "<Array ComplexPair[3]>"
    assert pairs.shape == (3,)
    np.testing.assert_allclose(pairs.real, np.array([0.0, 1.0, 2.0]))
    np.testing.assert_allclose(pairs.imag, np.array([0.0, -1.0, -2.0]))
    np.testing.assert_allclose(pairs["real"], pairs.real)
    assert isinstance(pairs[1], cm.ComplexPair)
    assert pairs[1].real == 1.0
    assert pairs[1].imag == -1.0


def test_struct_array_initialization(cm):
    from_array_like = cm.ComplexPairArray([(1.0, 2.0), (3.0, 4.0)])
    assert isinstance(from_array_like[0], cm.ComplexPair)
    np.testing.assert_allclose(from_array_like.real, [1.0, 3.0])
    np.testing.assert_allclose(from_array_like.imag, [2.0, 4.0])

    from_fields = cm.ComplexPairArray(real=[5.0, 6.0], imag=-1.0)
    np.testing.assert_allclose(from_fields.real, [5.0, 6.0])
    np.testing.assert_allclose(from_fields.imag, [-1.0, -1.0])


def test_struct_array_member(cm):
    p = cm.Particle()
    p.pos = [1.0, 2.0, 3.0]
    p.vel = [0.5, 0.5, 0.5]
    np.testing.assert_allclose(p.pos, np.array([1.0, 2.0, 3.0]))
    np.testing.assert_allclose(p.vel, np.array([0.5, 0.5, 0.5]))
    p.pos[:] *= 2.0
    np.testing.assert_allclose(p.pos, np.array([2.0, 4.0, 6.0]))


def test_struct_array_argument(cm):
    particles = cm.Particle.zeros(2)
    particles["pos"][0] = [0.0, 0.0, 0.0]
    particles["vel"][0] = [1.0, 0.0, 0.0]
    particles["pos"][1] = [0.0, 0.0, 0.0]
    particles["vel"][1] = [0.0, 2.0, 0.0]
    ke = cm.kinetic_energy(particles, len_p=len(particles))
    assert np.isclose(ke, 0.5 * (1.0 + 4.0))


def test_two_output_arrays(cm):
    data = np.array([0, 1, 2, 3, 4, 5], dtype=np.float64)
    out1, out2 = cm.split_vectors(data)
    np.testing.assert_allclose(out1, np.array([0, 2, 4], dtype=np.float64))
    np.testing.assert_allclose(out2, np.array([1, 3, 5], dtype=np.float64))


def test_merge_sorted(cm):
    a = np.array([1.0, 2.0, 2.0, 4.0], dtype=np.float64)
    b = np.array([2.0, 3.0], dtype=np.float64)
    out_len = np.zeros(1, dtype=np.int32)
    merged, out_len_val = cm.merge_sorted(a, b, out_len=out_len)
    np.testing.assert_allclose(merged, np.array([1.0, 2.0, 3.0, 4.0]))
    assert out_len_val == 4


def test_owned_array(cm):
    arr, n = cm.alloc_random_array()
    assert arr.shape[0] == n
    assert n >= 3
    assert np.all(arr[:n] >= 1.0)


def test_struct_output_array(cm):
    n = 4
    particles = cm.Particle.zeros(n)
    res = cm.make_particles(3.0, out_p=particles, len_p=n)
    assert res is None
    np.testing.assert_allclose(particles["pos"], np.ones((n, 3)))
    np.testing.assert_allclose(particles["vel"], np.array([[3.0, 0.0, 0.0]] * n))


def test_struct_output_array_auto_return(cm):
    n = 3
    particles = cm.make_particles(2.5, len_p=n)
    np.testing.assert_allclose(particles["pos"], np.ones((n, 3)))
    np.testing.assert_allclose(particles["vel"], np.array([[2.5, 0.0, 0.0]] * n))
